Use Github workflows

NeroAzure opened the pull request:
This PR adds workflow files to test, build and push the images to docker hub.

pehbehbeh commented on .github/workflows/testImageBuild.yml:
```diff
@@ -0,0 +1,14 @@
+name: Build test image
```
```suggestion
name: Test image build
```

pehbehbeh commented on .github/workflows/testImageBuild.yml:
```diff
@@ -0,0 +1,14 @@
+name: Build test image
+
+on: [pull_request]
```
Should we also run this for `develop` (or even `main`) to see if the changes affect anything else?

NeroAzure replied:
Now that you mention this, it would really make sense, on develop, but also on main (in case of hotfixes etc.). I will on this.

pehbehbeh commented on .github/workflows/stableImageBuild.yml:
```diff
@@ -0,0 +1,76 @@
+        name: Set up Docker Buildx
+        uses: docker/setup-buildx-action@v1
+      -
+        name: Login to DockerHub
```
```suggestion
        name: Login to Docker Hub
```

pehbehbeh commented on .github/workflows/edgeImageBuild.yml:
```diff
@@ -0,0 +1,46 @@
+          push: true
+          target: runtime
+          tags: |
+            sourceboat/docker-phoenix:edge-runtime
```
I would rename the file from `edgeImageBuild` to `buildEdgeImage`. Similar to `testImageBuild` the verb (`build` or `test`) of what the action does is always in front. This also groups similar actions together in the the filetree.

pehbehbeh commented on .github/workflows/stableImageBuild.yml:
```diff
@@ -0,0 +1,76 @@
+            sourceboat/docker-phoenix
+          flavour: |
+            latest=false
+          # generate Docker tags based on the following events/attributes
```
Personally I do not need any documentation in these files. I think it is pretty clear already what the code does.

NeroAzure commented:
There is still one thing to note: The test and build workflows would run in parallel, there is no "easy" way to define dependencies between different workflows, we either have to define separate workflows which include all necessary jobs for the trigger or do something with githubs `workflow_run` api, which I do not understand entirely and if it is even feasable for this application.

- https://github.community/t/how-do-i-specify-job-dependency-running-in-another-workflow/16482/8
- https://github.community/t/depend-on-another-workflow/16311/3

pehbehbeh commented:
> The test and build workflows would run in parallel [...]

So this could lead to a published image, which contains an error. That would be pretty bad. Maybe we should merge the build and test jobs into a single workflow and use `needs`?

https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#jobsjob_idneeds

NeroAzure commented:
> Maybe we should merge the build and test jobs into a single workflow and use needs?

Yes that is what I was suggesting with the separate workflows for each trigger, where triggers are the events triggering the workflow.

- develop: test;build;push (edge)
- main: test;build;push (stable)
- pr: test

NeroAzure commented:
I have updated the workflows to include the tests. I also changed the names, to `Edge`, `Stable` and `TestPRs` to reflect the end results of the workflows.

pehbehbeh commented:
LGTM 🥳
